protocol/contracts/voting/MasterCandidate.sol

```diff
@@ -0,0 +1,191 @@
+        nominationFeeToken = _nominationFeeToken;
+        nominationFee = _nominationFee;
+    }
+
+    /**
+     * @dev Change the nomination fee
+     * 
+     * Designed to be low, but a defense in depth to stop peeps putting up a nomination
+     * for the lolz.
+     */
+    function setNominationFee(uint256 _nominationFee) external onlyOwner {
+        nominationFee = _nominationFee;
+    }
+
+    /**
+     * @dev periodically update the merkle tree root to match discord membership.
+     * NOTE: If a user has already minted their 'master candidate' NFT, this doesn't
+     * remove it for them.
+     * 
+     * The voting UI only shows candidates in the intersection of the set
+     * MasterCandidate NFT and Discord Users.
+     */
+    function updateMerkleRoot(bytes32 _merkleRoot) external onlyOwner {
+        merkleRoot = _merkleRoot;
+    }
+
+    function nominate(
+        address account,
+        uint256 discordId,
+        string memory discordHandle,
+        string memory twitterHandle,
+        bytes32[] calldata merkleProof
+    ) external {
+        require(account != address(0), "MasterCandidate: Address cannot be 0x0");
+        require(MerkleProof.verify(merkleProof, merkleRoot, keccak256(abi.encode(account, discordId.toString()))), "MasterCandidate: candidate not in temple discord");
+
+        SafeERC20.safeTransferFrom(nominationFeeToken, account, address(this), nominationFee);
+        _safeMint(account, discordId);
+
+        Candidate storage candidate = candidates[discordId];
+        candidate.discordId = discordId;
+        candidate.discordHandle = discordHandle;
+        candidate.twitterHandle = twitterHandle;
+
+        emit UpdateMetadata(msg.sender, discordId, discordHandle, twitterHandle);
+    }
+
+    function updateMetadata(
+        uint256 discordId,
+        string memory discordHandle,
+        string memory twitterHandle
+    ) external {
+        require(ownerOf(discordId) == msg.sender, "Only the candidate can update metadata");
+        
+        Candidate storage candidate = candidates[discordId];
+        candidate.discordHandle = discordHandle;
+        candidate.twitterHandle = twitterHandle;
+
+        emit UpdateMetadata(msg.sender, discordId, discordHandle, twitterHandle);
+    }
+
+    function toggleEndorsement(uint256 discordId) external {
+        _toggleEndorsement(msg.sender, discordId);
+    }
+
+    function toggleEndorsements(uint256[] memory discordIds) external {
+        _toggleEndorsements(msg.sender, discordIds);
+    }
+
+    /**
+     * @notice Withdraw for another user (gasless for the vault token holder)
+     * (assuming the owner has given authority for the caller to act on their behalf)
+     *
+     * @dev amount is explicit, to allow use case of partial vault withdrawals
+     */
+    function toggleEndorsementsFor(address account, uint256[] memory discordIds, uint256 deadline, uint8 v, bytes32 r, bytes32 s) public {
```

Comment:
This is for gasless voting.

Our server can act as a relay that posts the transactions onto gnosis (in fact, anyone can run a relay. We should open source the code :p).

I'd imagine in the dapp, you can vote by bridging onto gnosis and doing the transactions yourself, or posting a signed transaction to a relay.

I *hope* this will loved the barrier to voting.